diff -puNr qt-everywhere-opensource-src-4.6.1.orig/tools/qdoc3/qdoc3.pro qt-everywhere-opensource-src-4.6.1/tools/qdoc3/qdoc3.pro
--- qt-everywhere-opensource-src-4.6.1.orig/tools/qdoc3/qdoc3.pro	2010-01-15 18:01:29.000000000 +0100
+++ qt-everywhere-opensource-src-4.6.1/tools/qdoc3/qdoc3.pro	2010-01-20 01:27:33.000000000 +0100
@@ -105,5 +105,6 @@ SOURCES += apigenerator.cpp \
            webxmlgenerator.cpp \
 	   yyindent.cpp
 
+TARGET = qdoc3
 target.path = $$[QT_INSTALL_BINS]
 INSTALLS += target
diff -puNr qt-everywhere-opensource-src-4.6.1.orig/tools/tools.pro qt-everywhere-opensource-src-4.6.1/tools/tools.pro
--- qt-everywhere-opensource-src-4.6.1.orig/tools/tools.pro	2010-01-15 18:01:30.000000000 +0100
+++ qt-everywhere-opensource-src-4.6.1/tools/tools.pro	2010-01-20 01:29:10.000000000 +0100
@@ -14,6 +14,7 @@ no-png {
          SUBDIRS += designer
      }
      SUBDIRS     += linguist
+     SUBDIRS     += qdoc3
      symbian: SUBDIRS = designer
      wince*: SUBDIRS = qtestlib designer
      unix:!mac:!embedded:contains(QT_CONFIG, qt3support):SUBDIRS += qtconfig
diff -puNr qt-everywhere-opensource-src-4.6.1.orig/tools/tools.pro~ qt-everywhere-opensource-src-4.6.1/tools/tools.pro~
